plugin/trino-sqlserver/src/main/java/io/trino/plugin/sqlserver/SqlServerClient.java

```diff
@@ -470,6 +499,46 @@ public void setNull(PreparedStatement statement, int index)
         };
     }
 
+    public static ColumnMapping sqlServerTimestampColumnMapping(TimestampType type)
+    {
+        if (type.getPrecision() <= MAX_SHORT_PRECISION) {
+            return timestampColumnMapping(type);
+        }
+        return ColumnMapping.objectMapping(
+                type,
+                longTimestampReadFunction(type),
+                longTimestampWriteFunction());
+    }
+
+    private static ObjectReadFunction longTimestampReadFunction(TimestampType type)
+    {
+        return ObjectReadFunction.of(
+                LongTimestamp.class,
+                (resultSet, columnIndex) -> {
+                    LocalDateTime localDateTime = resultSet.getObject(columnIndex, LocalDateTime.class);
+                    Instant instant = localDateTime.atZone(UTC).toInstant();
+                    long epochMicros = instant.getEpochSecond() * MICROSECONDS_PER_SECOND + floorDiv(instant.getNano(), NANOSECONDS_PER_MICROSECOND);
+                    int picosOfMicro = (instant.getNano() % NANOSECONDS_PER_MICROSECOND) * PICOSECONDS_PER_NANOSECOND;
+                    verifyLongTimestamp(picosOfMicro, type.getPrecision());
+                    return new LongTimestamp(epochMicros, picosOfMicro);
+                });
+    }
+
+    private static ObjectWriteFunction longTimestampWriteFunction()
+    {
+        return ObjectWriteFunction.of(
+                LongTimestamp.class,
+                (statement, index, value) -> {
+                    long epochSeconds = floorDiv(value.getEpochMicros(), MICROSECONDS_PER_SECOND);
+                    int microOfSecond = floorMod(value.getEpochMicros(), MICROSECONDS_PER_SECOND);
+                    int nanosOfSecond = (microOfSecond * NANOSECONDS_PER_MICROSECOND) +
```

Comment:
Just to confirm, values with precision > 7 are getting truncated?

Reply:
They should round. Do we have a test with CTAS in Trino with TIMESTAMP with precision>7 ?

Reply:
I'm not sure where rounding happens in Trino exactly. These tests prove that higher precision values are rounded:
https://github.com/trinodb/trino/pull/6654/files#diff-9ab27de6e7745b44111eb9dd53b1e9a446f9d02b29a8a6f20d81c8eae430b472R170

Reply:
This was addressed in another thread.